test/projection.jl

```diff
@@ -1,240 +1,249 @@
+        @test pstri(pstri(reshape(1:9,3,3))) == pstri(reshape(1:9,3,3))
+        @test pstri(rand(ComplexF32, 3, 3)) isa SymTridiagonal{Float64}
+        stri = SymTridiagonal(Symmetric(rand(3,3) .+ im))
+        @test pstri(stri) == real(stri)
+        @test_throws DimensionMismatch pstri(rand(ComplexF32, 3, 2))
+
+        ptri = ProjectTo(Tridiagonal(rand(3,3)))
+        @test ptri(reshape(1:9,3,3)) == [1.0 4.0 0.0; 2.0 5.0 8.0; 0.0 6.0 9.0]
+        @test ptri(ptri(reshape(1:9,3,3))) == ptri(reshape(1:9,3,3))
+        @test ptri(rand(ComplexF32, 3, 3)) isa Tridiagonal{Float64}
+        @test_throws DimensionMismatch ptri(rand(ComplexF32, 3, 2))
+
+        # an experiment with allowing subspaces which aren't subtypes
+        @test psymm(pdiag(rand(ComplexF32, 3, 3))) isa Diagonal{Float64}
+    end
 
-        x = zeros(2)
-        @test x == ProjectTo(x)(@thunk(ZeroTangent()))
+    @testset "SparseArrays" begin
+        # vector
+        v = sprand(30, 0.3)
+        pv = ProjectTo(v)
 
-        x = zeros(Float32, 2, 2)
-        @test x == ProjectTo(x)([0.0 0; 0 0])
+        @test pv(v) == v
+        @test pv(v .* (1+im)) ≈ v
+        o = pv(ones(Int, 30, 1))
+        @test nnz(o) == nnz(v)
 
-        x = [1.0 0; 0 4]
-        @test x == ProjectTo(x)(Diagonal([1.0, 4]))
+        # matrix
+        m = sprand(10, 10, 0.3)
+        pm = ProjectTo(m)
 
-        # to a array of structs
-        x = [Fred(0.0), Fred(0.0)]
-        @test x == ProjectTo(x)([Fred(0.0), Fred(0.0)])
-        @test x == ProjectTo(x)([ZeroTangent(), ZeroTangent()])
-        @test x == ProjectTo(x)([ZeroTangent(), @thunk(Fred(0.0))])
-        @test x == ProjectTo(x)(ZeroTangent())
-        @test x == ProjectTo(x)(@thunk(ZeroTangent()))
+        @test pm(m) == m
+        @test pm(m .* (1+im)) ≈ m
+        om = pm(ones(Int, 10, 10))
+        @test nnz(om) == nnz(m)
 
-        x = [Fred(1.0) Fred(0.0); Fred(0.0) Fred(4.0)]
-        @test x == ProjectTo(x)(Diagonal([Fred(1.0), Fred(4.0)]))
+        @test_throws DimensionMismatch pv(ones(Int, 1, 30))
+        @test_throws DimensionMismatch pm(ones(Int, 5, 20))
     end
 
-    @testset "To Array of Arrays" begin
-        # inner arrays have same type but different sizes
-        x = [[1.0, 2.0, 3.0], [4.0, 5.0]]
-        @test x == ProjectTo(x)(x)
-        @test x == ProjectTo(x)([[1.0 + 2im, 2.0, 3.0], [4.0 + 2im, 5.0]])
+    @testset "OffsetArrays" begin
+        # While there is no code for this, the rule that it checks axes(x) == axes(dx) else
+        # reshape means that it restores offsets. (It throws an error on nontrivial size mismatch.)
 
-        # This makes sure we don't fall for https://github.com/JuliaLang/julia/issues/38064
-        @test [[0.0, 0.0, 0.0], [0.0, 0.0]] == ProjectTo(x)(ZeroTangent())
-    end
+        poffv = ProjectTo(OffsetArray(rand(3), 0:2))
+        @test axes(poffv([1,2,3])) == (0:2,)
+        @test axes(poffv(hcat([1,2,3]))) == (0:2,)
```

Comment:
Reshaping has a nice side-effect that it restores offsets, which Zygote often loses.

Since it checks axes `==` but not `===` before reshaping, it does not at present restore StaticArrays. But perhaps those should be handled separately.